File: le/lib/q_0121_max_profit.ex
defmodule Solution0121 do
  @spec max_profit(prices :: [integer]) :: integer
  def max_profit(prices) do
    prices
    |> Enum.reduce({0, hd(prices)}, fn price, {max_profit, min_price} ->
      new_max_profit = max(max_profit, price - min_price)
      new_min_price = min(min_price, price)
      {new_max_profit, new_min_price}
    end)
    |> elem(0)
  end
end

File: lt/src/q0121MaxProfit.ts
export function maxProfit(prices: number[]): number {
    if (prices.length == 1) {
        return 0;
    }
    let maxProfit = 0;
    let [l, r] = [0, 1];

    while (r < prices.length) {
        if (prices[r] - prices[l] > 0) {
            maxProfit = Math.max(maxProfit, prices[r] - prices[l]);
            r += 1;
        } else {
            l = r;
            r += 1;
        }
    }

    return maxProfit;
}
